plugin/pulpcore/plugin/download/asyncio/group.py

```diff
@@ -0,0 +1,190 @@
+        >>>    yield (id, [DeferredArtifact(url=url_a), DeferredArtifact(url=url_b)])
+        >>>
+        >>> downloader = GroupDownloader(importer)
+        >>> downloader.schedule_from_iterator(group_generator)
+        >>> for id, group_results in downloader:
+        >>>     print(id)  # id is set to 'my_id'
+        >>>     # group_results is a dict of DownloadResult objects keyed by DeferredArtifact
+    """
+
+    def __init__(self, importer, download_class=HttpDownloader):
+        """
+        Args:
+            importer (pulpcore.plugin.models.Importer): The importer used to configure downloaders
+                with.
+            download_class (pulpcore.plugin.download.HttpDownloader): The asyncio-aware downloader
+                class the GroupDownloader will use when instantiating downloaders. This option
+                allows the GroupDownloader to work with custom downloaders.
+        """
+        self.importer = importer
+
+        self.download_factory = DownloadFactory(self.importer, download_class=download_class)
+        self.group_iterator = None
+        self.downloads_not_done = set()
+        self.groups_not_done = []
+        self.urls = defaultdict(list)  # dict with url as the key and a lists of Groups as the value
+        self.loop = asyncio.get_event_loop()
+
+    def schedule_from_iterator(self, group_iterator, parallel_group_limit=50):
+        """
+        Schedule groups to be downloaded using an iterator provided by the user.
+
+        When scheduling groups from the iterator, the `parallel_group_limit` argument defines the
+        number of groups that are handled in parallel. Once one group completes and is returned to
+        the user, another one is scheduled which maintains the number of parallel groups until the
+        iterator is exhausted.
+
+        When the group_iterator is a generator, this technique limits the effective memory used by
+        the DeferredDownload objects and their associated downloaders.
+
+        Additional calls to this will correctly schedule additional iterators, but only the first
+        call will specify the `parallel_group_limit`.
+
+        Args:
+            group_iterator (iterable): An iterable of tuples (id, [DeferredArtifact]). Each tuple is
+                one group with `id` being returned to the caller when all `DeferredArtifact.url`
+                downloads are complete.
+            parallel_group_limit (int): The number of groups to be handled in parallel at any time.
+        """
+        if self.group_iterator:
+            # This is not the first call to schedule an iterable
+            self.group_iterator = chain(self.group_iterator, group_iterator)
+            return
+
+        self.group_iterator = group_iterator
+        for i in range(parallel_group_limit):
+            with suppress(StopIteration):
+                self.schedule_group(**next(self.group_iterator))
+
+    def schedule_group(self, id, deferred_artifacts):
```

Comment:
Suggest this signature be `schedule_group(group)` for better symmetry with `schedule_from_iterator()`.

Reply:
This is a good idea. Fixed in next push.